lib/charms/kubernetes_charm_libraries/v0/volumes.py

```diff
@@ -0,0 +1,415 @@
+        if not statefulset_spec.template.spec.volumes:
+            return False
+        return all(
+            [
+                requested_volume.volume in statefulset_spec.template.spec.volumes
+                for requested_volume in requested_volumes
+            ]
+        )
+
+    def _pod_contains_requested_volumes(
+        self,
+        container_name: str,
+        requested_volumes: Iterable[RequestedVolume],
+        pod: Union[PodTemplateSpec, Pod],
+    ) -> bool:
+        """Returns whether a pod is patched with requested volumes.
+
+        Args:
+            container_name: Container name
+            requested_volumes: Iterable of requested volumes
+            pod: Kubernetes pod object.
+
+        Returns:
+            bool
+        """
+        return self._pod_volumemounts_contain_requested_volumes(
+            containers=pod.spec.containers,
+            container_name=container_name,
+            requested_volumes=requested_volumes,
+        )
+
+    @staticmethod
+    def _pod_volumemounts_contain_requested_volumes(
+        containers: Iterable[Container],
+        container_name: str,
+        requested_volumes: Iterable[RequestedVolume],
+    ) -> bool:
+        """Returns whether container spec contains the expected requested volumes mounts.
+
+        Args:
+            containers: Iterable of Containers
+            container_name: Container name
+            requested_volumes: Requested volumes we expect to be set
+
+        Returns:
+            bool
+        """
+        container = next(container for container in containers if container.name == container_name)
+        return all(
+            [
+                requested_volume.volume_mount in container.volumeMounts
+                for requested_volume in requested_volumes
+            ]
+        )
+
+    def patch_volumes(
+        self,
+        statefulset_name: str,
+        requested_volumes: Iterable[RequestedVolume],
+        container_name: str,
+    ) -> None:
+        """Patches a statefulset with requested volumes.
+
+        Args:
+            statefulset_name: Statefulset name
+            requested_volumes: Iterable of requested volumes
+            container_name: Container name
+        """
+        if not requested_volumes:
+            logger.info("No requested volumes were provided")
+            return
+        try:
+            statefulset = self.client.get(
+                res=StatefulSet, name=statefulset_name, namespace=self.namespace
+            )
+        except ApiError:
+            raise KubernetesRequestedVolumesError(f"Could not get statefulset {statefulset_name}")
+        container = Container(
+            name=container_name,
+            volumeMounts=[requested_volume.volume_mount for requested_volume in requested_volumes],
+            resources=ResourceRequirements(limits={}, requests={}),
+        )
+        if any(
+            [
+                requested_volume.volume.emptyDir.medium == "HugePages"
```

Comment:
I would extract this into a  _is_huge_pages_volume_requested or something similar to improve readability. It can also be reused in the remove method